python/pyspark/errors/exceptions/base.py

```diff
@@ -449,3 +449,35 @@ def summary(self) -> str:
         Summary of the exception cause.
         """
         ...
+
+
+T = TypeVar("T", bound=PySparkException)
+
+
+def recover_python_exception(e: T) -> T:
+    """
+    Recover Python exception stack trace.
+
+    Many JVM exceptions types may wrap Python exceptions. For example:
+    - UDFs can cause PythonException
+    - UDTFs and Data Sources can cause AnalysisException
+    """
+    python_exception_header = "Traceback (most recent call last):"
+    try:
+        from pyspark.errors.exceptions.tblib import Traceback
+
+        message = str(e)
+        start = message.find(python_exception_header)
+        if start == -1:
+            # No Python exception found
+            return e
+
+        # The message contains a Python exception. Parse it to use it as the exception's traceback.
+        # This allows richer error messages, for example showing line content in Python UDF.
+        python_exception_string = message[start:]
+        tb = Traceback.from_string(python_exception_string)
+        tb.populate_linecache()
+        return e.with_traceback(tb.as_traceback())
+    except Exception:
```

Comment:
I think we need a configuration, or at least environment variable if it's difficult to add a configuration here (for the cases when exceptions are thrown without JVM). Parsing exceptions can potentially cause a lot of performance overhead, e.g., if users are relaying on a lot of exceptions.

In addition, it would have to be `BaseException` if you absolutely want to catch all the exceptions.

Reply:
> Parsing exceptions can potentially cause a lot of performance overhead

Parsing should take time linear to the size of the error message so I don't worry much about performance. Let me do a quick benchmark to validate this.

> In addition, it would have to be BaseException if you absolutely want to catch all the exceptions.

Good point :+1: (although I don't think it could possibly throw stuff like `KeyboardInterrupt` or `SystemExit`, but let's be safe)

Reply:
Here's a quick benchmark with different traceback sizes (1, 10, 100, 500, 1000)
https://perfpy.com/985

| Benchmark | Runtime |
| --------- | ------- |
| 1 frame  | 297 us  |
| 10 frames  | 335 us  |
| 100 frames    | 2.41 ms |
| 500 frames  | 12.4 ms |
| 1000 frames | 28.7 ms |
| 1000 frames, failed to parse | 3.13 ms |

The default recursion limit is 1000 so tracebacks with more than 1000 frames should be unlikely.

Also, the results I see when I benchmark locally is around 1/5 of the above. Therefore performance shouldn't be a concern.